### ISSUE TITLE
Function Apps using Durable Task in Isolated worker: Inner exception being lost

### ISSUE THREAD

**Issue by wwalkley** [Azure/azure-functions-durable-extension#2697]
### Description

After migrating to the Isolated worker process, we've encountered an issue with handling exceptions. The `FunctionFailedException` is no longer accessible in the latest worker packages. Despite extensive searching, there's no clear documentation or online resources addressing this issue.

The documentation states that exceptions thrown in activity functions should be marshaled back to the orchestrator function as `FunctionFailedException`. However, in our case, they are propagating up as `TaskFailedException`'s. This behavior differs from the expected outcome, as outlined here: #https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-error-handling?tabs=csharp-isolated#errors-in-activity-functions

> Any exception that is thrown in an activity function is marshaled back to the orchestrator function and thrown as a `FunctionFailedException`. You can write error handling and compensation code that suits your needs in the orchestrator function.

### Expected behavior
The exception thrown contains the root exception and we have access to the `FunctionFailedException`

### Actual behavior
The exception thrown contains only the message of the root exception and is a `TaskFailedException`

### Relevant source code snippets

_Snippet in debug mode:_

![image](https://github.com/Azure/azure-functions-durable-extension/assets/29252552/cd4a474a-558b-4b8a-8d8a-42a2890c2043)

_Stacktrace:_

```csharp
Microsoft.DurableTask.TaskFailedException: Task 'GetCustomerIntegrationStatusActivity' (#6) failed with an unhandled exception: One or more errors occurred. (HealthCheckFailed Xero API health check failed. This could be transient on our side, transient on their side, or a more serious problem.)
 ---> DurableTask.Core.Exceptions.TaskFailedException: Exception of type 'DurableTask.Core.Exceptions.TaskFailedException' was thrown.
   at DurableTask.Core.TaskOrchestrationContext.ScheduleTaskInternal(String name, String version, String taskList, Type resultType, Object[] parameters) in /_/src/DurableTask.Core/TaskOrchestrationContext.cs:line 118
   at DurableTask.Core.TaskOrchestrationContext.ScheduleTaskToWorker[TResult](String name, String version, String taskList, Object[] parameters) in /_/src/DurableTask.Core/TaskOrchestrationContext.cs:line 89
   at DurableTask.Core.TaskOrchestrationContext.ScheduleTask[TResult](String name, String version, Object[] parameters) in /_/src/DurableTask.Core/TaskOrchestrationContext.cs:line 81
   at DurableTask.Core.RetryInterceptor`1.Invoke() in /_/src/DurableTask.Core/RetryInterceptor.cs:line 61
   at DurableTask.Core.RetryInterceptor`1.Invoke() in /_/src/DurableTask.Core/RetryInterceptor.cs:line 96
   at Microsoft.DurableTask.Worker.Shims.TaskOrchestrationContextWrapper.CallActivityAsync[T](TaskName name, Object input, TaskOptions options)
   --- End of inner exception stack trace ---
   at Microsoft.DurableTask.Worker.Shims.TaskOrchestrationContextWrapper.CallActivityAsync[T](TaskName name, Object input, TaskOptions options)
   at Rifleman.Fintegration.Functions.Helpers.SyncExecutors.FullSyncExecutor.GetCustomerSyncStatus() in C:\Work\Rifleman.Fintegration\Rifleman.Fintegration\Rifleman.Fintegration.Functions\Helpers\SyncExecutors\FullSyncExecutor.cs:line 95
   at Rifleman.Fintegration.Functions.Helpers.SyncExecutors.FullSyncExecutor.GetCustomerSyncStatus() in C:\Work\Rifleman.Fintegration\Rifleman.Fintegration\Rifleman.Fintegration.Functions\Helpers\SyncExecutors\FullSyncExecutor.cs:line 129
   at Rifleman.Fintegration.Functions.Helpers.SyncExecutors.SyncExecutorBase.PerformPreSyncCheck() in C:\Work\Rifleman.Fintegration\Rifleman.Fintegration\Rifleman.Fintegration.Functions\Helpers\SyncExecutors\SyncExecutorBase.cs:line 79
```

### Known workarounds
To get around this, we caught `TaskFailedException`'s in our Orchestrators and used string comparisons against the `ex.FailureDetails.ErrorMessage` to handle certain expectations :cry:

### App Details

- **Azure Functions runtime version**: v4
- **Programming language used**: C#

```
<ItemGroup>
        <PackageReference Include="Microsoft.Azure.Core.NewtonsoftJson" Version="1.0.0"/>
        <PackageReference Include="Microsoft.Azure.Cosmos.Table" Version="1.0.8"/>
        <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.19.0"/>
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.DurableTask" Version="1.1.0"/>
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http" Version="3.1.0"/>
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore" Version="1.0.0"/>
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.ServiceBus" Version="5.14.1"/>
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Timer" Version="4.2.0"/>
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.15.1"/>
        <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.1.0"/>
        <PackageReference Include="Microsoft.Azure.Storage.Blob" Version="11.2.3"/>
        <PackageReference Include="Microsoft.Azure.Storage.Queue" Version="11.2.3"/>
        <PackageReference Include="Microsoft.DurableTask.Generators" Version="1.0.0-preview.1"/>
        <PackageReference Include="Microsoft.Extensions.Http" Version="6.0.0"/>
</ItemGroup>

```

Any guidance would be appreciated.

**Comment by cgillum** [comment]
Hi @imeya. As far as I can tell, this is by design, and I apologize that this was not documented clearly. [`TaskFailedException`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.durabletask.taskfailedexception) is the expected exception type which replaces `FunctionFailedException` from the .NET in-proc model.

Details about the root exception are in the [`TaskFailedException.FailureDetails`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.durabletask.taskfailuredetails) property. We are no longer serializing the original activity exception because we found this to be completely unreliable since many exception types (including some that are part of the framework) are not serializeable. Security is also somewhat of a concern since we don't have direct control over or visibility into what exception types are being serialized or deserialized.

Can you share a code snippet from your previous orchestration logic showing how you were handling exceptions previously in the .NET in-proc model?

**Comment by wwalkley** [comment]
@cgillum @imeya Thanks for getting back to us, we appreciate your cooperation as we figure out how to move forward. 

Following up on the documentation discrepancy regarding `TaskFailedException`. While your confirmation clarifies its expected behavior, the existing documentation inadvertently led us astray. It would be great to get the documentation updated. 
 
_Our code changes. Left is in-proc, Right is isolated worker_
![image](https://github.com/Azure/azure-functions-durable-extension/assets/29252552/f69ce5ba-50ba-406d-9ba8-4e94da26f560)

Relying on inner/base exceptions (highlighted in red) proved ineffective when moving to the Isolated worker model. This necessitated adding a catch block for `TaskFailedException`. However, we encountered a challenge: the `FailureDetails` property didn't provide the granular information needed for type-based exception handling. We've resorted to string comparison from the error message as a temporary workaround (see example). Attached are screenshots showcasing the revised error handling logic and the structure of FailureDetails. We'd love your input on how we can improve this further.

_Here is what the failure details look like:_
![image](https://github.com/Azure/azure-functions-durable-extension/assets/29252552/2dbf95f7-01e3-4de5-ae5e-ba4575740891)

Full details:
```
ErrorMessage: One or more errors occurred. (HealthCheckFailed Xero API health check failed. This could be transient on our side, transient on their side, or a more serious problem.)
ErrorType: (unknown)
InnerFailure: null
ExceptionType: null
StackTrace:
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionInvoker`2.<>c.<InvokeAsync>b__6_0(Task`1 t) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionInvoker.cs:line 32
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 45
   at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
   at Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.FunctionsHttpProxyingMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\extensions\Worker.Extensions.Http.AspNetCore\src\FunctionsMiddleware\FunctionsHttpProxyingMiddleware.cs:line 34
   at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77
   at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler.InvokeAsync(InvocationRequest request) in D:\a\_work\1\s\src\DotNetWorker.Grpc\Handlers\InvocationHandler.cs:line 88


```
_Custom exception from example:_
```
[Serializable]
public class HealthCheckException : Exception
{
	public HealthCheckException( ) { }
	public HealthCheckException( string message ) : base( $"{Constants.ErrorMessage.HealthCheckFailed} {message}" ) { }
	public HealthCheckException( string message, Exception innerException ) : base( $"{Constants.ErrorMessage.HealthCheckFailed} {message}", innerException ) { }
	protected HealthCheckException( SerializationInfo info, StreamingContext context ) : base( info, context ) { }
}
```

**Comment by cgillum** [comment]
@wwalkley thanks for the detailed information. Try the following as an alternative way to do the error handling, which will more closely resemble what you had before:

```csharp
try
{
    // ...
}
catch (TaskFailedException ex) when (ex.FailureDetails.IsCausedBy<ExternalAccessRevokedException>())
{
    // ...
}
catch (TaskFailedException ex) when (ex.FailureDetails.IsCausedBy<HealthCheckException>())
{
    // ...
}
```

**Comment by wwalkley** [comment]
Nice method! Unfortunately it returned `false` due to `ExceptionType` being `null` and `ErrorType` being `unknown`.

![image](https://github.com/Azure/azure-functions-durable-extension/assets/29252552/2646c6af-6743-45c2-9bf3-b7285fba88cd)

Any ideas?

**Comment by jviau** [comment]
I believe this is an issue of us not capturing inner exceptions in failure details. We should move this to microsoft/durabletask-dotnet as the issue lies there.

**Comment by wwalkley** [comment]
@jviau are the `microsoft/durabletask-dotnet` team across this issue?

**Comment by imeya** [comment]
@jviau thanks for help check the issue. how do we move, or you can help move this to microsoft/durabletask-dotnet? Thanks very much.

**Comment by lilyjma** [comment]
@wwalkley @imeya - looks like this issue can't be moved because this repo is in a different organization (Azure) as the other one (microsoft). _But_, we're the team that's maintaining both repos so this issue is under our radar.

**Comment by imeya** [comment]
thanks @lilyjma for looking into the issue. any ETA for this issue? Thanks.

**Comment by EmilDamsbo** [comment]
The documentation page here should also mention that `TaskFailedException` is the replacement: https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-error-handling?tabs=csharp-isolated#errors-in-activity-functions

This was the only place I could search up that mentioned the change.

**Comment by wwalkley** [comment]
@lilyjma @imeya . Appreciate the update! Glad the team's looking into this.

**Comment by yatiac** [comment]
@jviau @lilyjma do you have any ETA on this one? or an alternative/work around? Thanks in advance

**Comment by miniGweek** [comment]
Looking for a solution or workaround for this too.

**Comment by cgillum** [comment]
I created a local repro and can confirm that the problem is as described [here](https://github.com/Azure/azure-functions-durable-extension/issues/2697#issuecomment-1861776982). The issue seems to be that the code for loading the exception type only works for certain built-in exception types (like `System.ApplicationException`, which is what was tested) but not for things like custom exception types.

I will look into creating a fix for this.

**Comment by cgillum** [comment]
This will be fixed in the next release of the .NET Isolated nuget package (Microsoft.Azure.Functions.Worker.Extensions.DurableTask).

**Comment by milkyware** [comment]
Hi, one thing I've encountered is that with the **EnableUserCodeException** [worker option enabled](https://learn.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-process-guide?tabs=windows), the ErrorType is unknown

![image](https://github.com/Azure/azure-functions-durable-extension/assets/24723886/56a0e26b-2136-42b4-8ce6-6f8ef894e35f)

Is the updated package expected to fix this scenario as well, or is it a separate issue?

This is using the 1.2.0-rc.1 version of Microsoft.Azure.Functions.Worker.Extensions.DurableTask

**Comment by wwalkley** [comment]
I apologize for bringing this up again, but we recently upgraded the DurableTask package to version `1.1.2`. Unfortunately, the issue we were encountering hasn't been resolved.

The error message we're receiving is still categorized as `unknown` which prevents us from understanding the root cause of the failure. Furthermore, the inner failure is null.

_What we are seeing:_
![image](https://github.com/Azure/azure-functions-durable-extension/assets/29252552/9ca31765-23a3-452a-a5e3-e4e830144782)

Due to the unknown error type, we're unable to leverage the recommended, more specific error handling strategies from this thread. Instead, we're forced to rely on a string comparison against the `ErrorMessage`. 

_These errors aren't caught and instead it falls through to the general Catch at the bottom:_
![image](https://github.com/Azure/azure-functions-durable-extension/assets/29252552/56b34f2f-9250-4b8f-bf61-bcbbb0d571d3)

Any ideas or suggestions would be greatly appreciated, thanks.

**Comment by cgillum** [comment]
Hi @wwalkley, if you haven't done so already, can you try upgrading to the latest version of the Durable nuget package ([v1.1.7](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.Extensions.DurableTask/1.1.7) at the time of writing)? The issue you're running into was fixed starting with the [v1.2.3](https://github.com/microsoft/durabletask-dotnet/releases/tag/v1.2.3) release of the core Durable Task SDK, which is transitively included in the v1.1.7 extension release (sorry that the version numbers are a bit confusing).
